### ISSUE TITLE
GCP KMS - SCT support

### ISSUE THREAD

**Issue by mykaul** [scylladb/scylla-cluster-tests#7997]
GCP KMS is not covered in SCT today (AFAIK) and is coming in 2024.2. We need to add, similar to AWS KMS, support to it.

**Comment by rayakurl** [comment]
@lersek @elcallio is there an ETA for this task?

**Comment by lersek** [comment]
@rayakurl this is at the bottom of my TODO list. I'm a beginner; I've been working through the "more approachable" items near the top of my TODO list (which are still hard).

I'll get to this one (I'm trying *not* to insert further (smaller) items near the top of my list), but I have no ETA. I don't know nearly enough to attack this problem at once, especially not concurrently with other assignments. Seasoned developers can multi-task well (speaking from experience), but it depends on lexical knowledge, and I'm not there yet; I have minimal background in scylla, for the time being.

**Comment by rjeczalik** [comment]
FWIW This is Terraform template for GCP KMS keys that cloud uses in its automation:

https://github.com/scylladb/siren-byok/blob/main/services/key-service/server/parts/gcp/key.yaml

I tested `2024.2.0~rc2-0.20240904.4c26004e5311` manually and it looks like it works.

For the GCP KeyRing and CryptoKey we are using defaults, and I am wondering if defaults are good enough or do we need to somehow harden them?

**Comment by lersek** [comment]
(Note to self: the one SCT user/developer guide I'm aware of is at https://docs.google.com/presentation/d/1QC9Tf4mVvs6yql7QmqUvJqABX2-mW2FN8JAvnRIMbQM/edit.)

**Comment by rjeczalik** [comment]
Also there is an aspect of CMK key rotation - AWS rotates underlying key material each 365 days, GCP - each 90 days.

It would be also good to cover if everything is sound when we force the rotation manually (under load). The answer - whether key rotation affects or does not affect the cluster serving data - will allow us to design cloud operations accordingly (a bit related issue here: [scylladb/siren-byok#449](https://github.com/scylladb/siren-byok/issues/449)):

a. if it does not affect (no performance issues reloading the DEK etc) - then we can relay on auto-rotation, unifying the intervals on both clouds.

b. if it affects - then we need to rotate manually in a maintenance window and this requires development.

@mykaul @gmizrahi FYI

**Comment by lersek** [comment]
With Barak, we've set up the permissions in the google cloud console as follows:
- The SCT service account now has `Service Account Token Creator` role (= impersonation right) for the preexistent `test1` user (who can use the preexistent test key for encryption and decryption). In this regard, the SCT service account is allowed to impersonate `test1` just like `test2` is allowed to do the same (used by scylla unit tests).
- The SCT service account also has direct `Cloud KMS CryptoKey Decrypter` and `Cloud KMS CryptoKey Encrypter` roles (similarly to the `test1` user) under the preexistent test key.

*Either* of these methods *should* suffice for the SCT service account to encrypt/decrypt data with the test key in GCE.

At first, I [chose](https://github.com/lersek/scylla-cluster-tests/commit/3139b6b2369d) the first approach; I hard-coded the following scylla config snippet in SCT:

~~~
diff --git a/configurations/gce-ear.yaml b/configurations/gce-ear.yaml
new file mode 100644
index 000000000000..e2b75926c4ad
--- /dev/null
+++ b/configurations/gce-ear.yaml
@@ -0,0 +1,13 @@
+scylla_encryption_options: >-
+  { 'cipher_algorithm' : 'AES/CBC/PKCS5Padding',
+    'secret_key_strength' : 128,
+    'key_provider': 'GcpKeyProviderFactory',
+    'gcp_host': 'gcp-1' }
+
+append_scylla_yaml:
+  gcp_hosts:
+    gcp-1:
+      gcp_project_id: scylla-kms-test
+      gcp_location: global
+      master_key: test_ring/test_key
+      gcp_impersonate_service_account: test1-REDACTED
~~~

However, the test failed. The `gcp` logger printed:
~~~
Looking up master key
Creating new key: AES:128
Attempting to load credentials from GCE
Resolved http://metadata.google.internal -> 169.254.169.254:80/
Sending  request to http://metadata.google.internal ([("Metadata-Flavor", "Google")]):
Got reponse 200: computeMetadata/
Resolved http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/token ->
169.254.169.254:80/computeMetadata/v1/instance/service-accounts/default/token
Sending  request to http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/token ([("Metadata-Flavor", "Google")]):
Got reponse 200: {"access_token":"REDACTED","expires_in":3599,"token_type":"Bearer"}
Resolved https://iamcredentials.googleapis.com/v1/projects/-/serviceAccounts/test1-REDACTED:generateAccessToken ->
172.217.203.95:443/v1/projects/-/serviceAccounts/test1-REDACTED:generateAccessToken
Sending application/json request to https://iamcredentials.googleapis.com/v1/projects/-/serviceAccounts/test1-REDACTED:generateAccessToken ([("Authorization",
"Bearer REDACTED")]): {"scope":["https://www.googleapis.com/auth/cloudkms"],"lifetime":"3600s"}

Got reponse 403: {
  "error": {
    "code": 403,
    "message": "Request had insufficient authentication scopes.",
    "status": "PERMISSION_DENIED",
    "details": [
      {
        "@type": "type.googleapis.com/google.rpc.ErrorInfo",
        "reason": "ACCESS_TOKEN_SCOPE_INSUFFICIENT",
        "domain": "googleapis.com",
        "metadata": {
          "method": "google.iam.credentials.v1.IAMCredentials.GenerateAccessToken",
          "service": "iamcredentials.googleapis.com"
        }
      }
    ]
  }
}
Got unexpected reponse (403 Forbidden)
X-Frame-Options: SAMEORIGIN
Server: scaffolding on HTTPServer2
X-Content-Type-Options: nosniff
X-XSS-Protection: 0
Date: Mon, 23 Dec 2024 11:22:48 GMT
Transfer-Encoding: chunked
Content-Type: application/json; charset=UTF-8
Accept-Ranges: none
Vary: X-Origin,Referer,Origin,Accept-Encoding
WWW-Authenticate: Bearer realm="https://accounts.google.com/", error="insufficient_scope", scope="https://www.googleapis.com/auth/iam
https://www.googleapis.com/auth/cloud-platform"
[same response as above]
~~~

What we see above is that the SCT account first gets a default access token for itself, and then tries to get an access token for `test1` through the default access token it has just obtained. However, google rejects the authorization attempt, *in spite of* the SCT service account having `Service Account Token Creator` permission under `test1`

And so the `init` logger ultimately prints:
~~~
Startup failed: std::_Nested_exception<encryption::permission_error> (Error refreshing credentials): seastar::httpd::unexpected_status_error (403 Forbidden)
~~~

In summary, this should be working, but it is not, and I don't know what other credential is missing.

**Comment by bhalevy** [comment]
@regevran @elcallio what's the status here?
What is the target milestone?

**Comment by regevran** [comment]
> @regevran @elcallio what's the status here?

It seems that @lersek initiated the work and it is almost ready.
@elcallio - please let me know if you need my support somehow to complete this.

**Comment by ndragazis** [comment]
My two cents:

The `ACCESS_TOKEN_SCOPE_INSUFFICIENT` means that the access token of the node's service account does not have sufficient permissions to hit the impersonation API. The docs clearly mention that one of the following scopes is required:
* `https://www.googleapis.com/auth/iam`
* `https://www.googleapis.com/auth/cloud-platform`

source: https://cloud.google.com/iam/docs/reference/credentials/rest/v1/projects.serviceAccounts/generateAccessToken#authorization-scopes

I expect that these scopes are missing in @lersek's demo, hence the error. If we could enter in one of those machines, we could actually inspect the token's scopes via the `tokeninfo` endpoint and confirm:
```
curl -H "Content-Type: application/x-www-form-urlencoded" -d "access_token=$(gcloud auth application-default print-access-token)" https://www.googleapis.com/oauth2/v1/tokeninfo
```

But why are these scopes missing?

Based on the SCT code (`gce_builder.py`), we are not setting any scopes explicitly when creating the instances (the API accepts `properties.serviceAccounts[].scopes[]`, see https://cloud.google.com/compute/docs/reference/rest/v1/instanceTemplates). This means that the default scopes are used: https://cloud.google.com/compute/docs/access/service-accounts#default_scopes

The ones we need are not in the defaults.

But frankly, I don't see the point of impersonation in testing. Since the SCT service account has already been configured with Cloud KMS access, we can use that directly.

**Comment by mykaul** [comment]
@regevran , @pehala - this is a gap we need to close.

**Comment by regevran** [comment]
@pehala - I think it is more on your end then mine.
I looked at the code trying to figure out if is closer to search-replace-tweak activity or more like refactor.
I suspect that the latter is the required work.

**Comment by pehala** [comment]
> [@pehala](https://github.com/pehala) - I think it is more on your end than mine. I looked at the code trying to figure out if is closer to search-replace-tweak activity or more like refactor. I suspect that the latter is the required work.

It is a full on feature for SCT, @Lakshmipathi is already working in it

**Comment by mykaul** [comment]
@pehala , @Lakshmipathi - I'll need a detailed timeline here. We wish to push it to the cloud, I'm very concerned without enough SCT coverage.

**Comment by pehala** [comment]
> [@pehala](https://github.com/pehala) , [@Lakshmipathi](https://github.com/Lakshmipathi) - I'll need a detailed timeline here. We wish to push it to the cloud, I'm very concerned without enough SCT coverage.

@Lakshmipathi verified with manual setup with a standard longevity run with nemesis that it indeed works (https://argus.scylladb.com/tests/scylla-cluster-tests/8a20925f-a025-46f2-abed-5d89da1fe83c/events), test failed only due to OVERSIZED_ALLOCATIONS. This gives us enough confidence in the feature for cloud release (email will be send shortly).

That said, the full on PR to SCT with enablement to all GCP test cases will still take a few more weeks

**Comment by mykaul** [comment]
We have a month to 2025.3 GA, so I set it to this milestone.

**Comment by mykaul** [comment]
@pehala - do we know where this stands?

**Comment by pehala** [comment]
> [@pehala](https://github.com/pehala) - do we know where this stands?

Yes, it is enabled by default and used
